src/api/artifact/controller.go

```diff
@@ -507,6 +512,48 @@ func (c *controller) RemoveLabel(ctx context.Context, artifactID int64, labelID
 	return c.labelMgr.RemoveFrom(ctx, labelID, artifactID)
 }
 
+func (c *controller) Walk(ctx context.Context, root *Artifact, walkFn func(*Artifact) error, option *Option) error {
+	queue := list.New()
+	queue.PushBack(root)
+
+	for queue.Len() > 0 {
+		elem := queue.Front()
+		queue.Remove(elem)
+
+		artifact := elem.Value.(*Artifact)
+		if err := walkFn(artifact); err != nil {
+			return err
+		}
+
+		if !artifact.HasChildren() {
```

Comment:
Whether the artifact contains children or not can be determined by the length of references. The `HasChildren` isn't needed?

Reply:
I'm not sure it's the correct way to determine whether the artifact contains children by checking the media type. But the checking the references must be

Reply:
I think we should check it by artifact media type, not the db query, and in artifact manager we should query references only when the media type is manifest index.